src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/DecompressionHandler.cs

```diff
@@ -80,7 +81,10 @@ internal override async ValueTask<HttpResponseMessage> SendAsync(HttpRequestMess
 
                 if (GZipEnabled && last == Gzip)
                 {
-                    response.Content = new GZipDecompressedContent(response.Content);
+                    if(await CheckContentSignatureAsync(await response.Content.ReadAsStreamAsync(cancellationToken), s_gzipSignature))
```

Comment:
I'm not convinced the concept here is something we should try to make work...

But that aside, this implementation won't work.  It's consuming multiple bytes from the stream.  If it is actually gzip, then having consumed the header the gzip lib won't see it as gzip because it lacks the header.  If it isn't actually gzip, then this will have consumed multiple bytes from the response stream.

Reply:
why do you think it is not useful?

i have the problem that a server is configured exactly the wrong way. since the class `DecompressionHandler` is internal sealed, i can't customize it either and have to re-implement / copy and customize all the code.

i can't think of any other solution.

btw, the position is resetet in CheckContentSignatureAsync, but as I said, it is untested and maybe I am also wrong

Reply:
> the position is resetet in CheckContentSignatureAsync

The streams involved here aren't seekable.

> why do you think it is not useful?

Because doing this correctly adds cost to workaround a server that's actively doing the wrong thing.  There are a myriad of ways it could be doing the wrong thing, and by returning the data without error, it's only working around this one very specific way the server is doing it wrong, and in doing so, passing along effectively corrupt data to the consumer.  What if the server said it was speaking gzip but sent down data as zlib or deflate?  What if the server said it was speaking gzip but didn't send the gzip header, only the compressed body and gzip footer?  Etc.

Reply:
> Because doing this correctly adds cost to workaround a server that's actively doing the wrong thing. There are a myriad of ways it could be doing the wrong thing, and by returning the data without error, it's only working around this one very specific way the server is doing it wrong, and in doing so, passing along effectively corrupt data to the consumer. What if the server said it was speaking gzip but sent down data as zlib or deflate? What if the server said it was speaking gzip but didn't send the gzip header, only the compressed body and gzip footer? Etc.

ok, i'm agree with this, but then we could simply return an meaningfull error instead of the wrong error message "The archive entry was compressed using an unsupported compression method." and returning something like "The request is announced to be compressed but isn't a valid gzip stream" and give the caller an ability to access the underlying stream.

somthing like `InvalidGzipStreamException:InvalidDataException(string mesage, Stream underlyingStream)`

Reply:
"The archive entry was compressed using an unsupported compression method." is wrong because *the stream is not compressed* and not *is compressed with an unsupported method*.

Reply:
Yes, the error message here (which comes from GZipStream) could be loosened to be less presumptive about why zlib reported Z_DATA_ERROR.
https://github.com/dotnet/runtime/blob/7b32a1ac009a1f892336a7515abc5b28c307f997/src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/Inflater.cs#L309-L310
cc: @dotnet/area-system-io-compression

Reply:
Did you have any suggestion to add a handler before:

https://github.com/dotnet/runtime/blob/ccfe21882e4a2206ce49cd5b32d3eb3cab3e530f/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs#L519

Reply:
> Did you have any suggestion to add a handler before:

I don't understand the question.  Can you elaborate?

Reply:
DelegatingHandlers i write are executed after the DecompressionHandler.
Because tha SocketsHttpHandler is sealed i also cant overwrite them.

I search for a way to execute a HttpMessageHandler before the DecompressionHandler is executed so i can remove the Content-Encoding header (if i detected the descriped issue) before the DecompressionHandler is executed.

The DocompressionHandler is added in front of the HandlerChain in SocketsHttpHandler.SetupHandlerChain()

https://github.com/dotnet/runtime/blob/ccfe21882e4a2206ce49cd5b32d3eb3cab3e530f/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs#L479

https://github.com/dotnet/runtime/blob/ccfe21882e4a2206ce49cd5b32d3eb3cab3e530f/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs#L519

witch is used by HttpClientHandler.

Reply:
> Because tha SocketsHttpHandler is sealed i also cant overwrite them.

You can set AutomaticDecompression to None and then do whatever decompression you want yourself in your wrapping handler.

Reply:
> > Because tha SocketsHttpHandler is sealed i also cant overwrite them.
> 
> You can set AutomaticDecompression to None and then do whatever decompression you want yourself in your wrapping handler.

I would also link this issue #24205. It would be helpful if the DecompressionHandler could be public in the future.

https://github.com/dotnet/runtime/issues/24205#issuecomment-540707116
> We can reconsider in future if there are more people who need it.